### ISSUE TITLE
Deploy-Eflow failure/could not track signiture to the microsoft root certification

### ISSUE THREAD

**Issue by elidevs** [Azure/iotedge-eflow#158]
**Question Summary**
Running into the error "could not track signiture to the microsoft root certification at line 5480" after running Deploy-Eflow on a Windows Server vm via PowerShell. Checking certificates on the VM, it only has Microsoft Defender and not the Microsoft Corporation one. Is the absence of the Microsoft Corporation certificate causing the signiture error? Am I missing any certificates on the VM?

**Detailed background**
Screenshot of error:
![MicrosoftTeams-image (1)](https://user-images.githubusercontent.com/16518452/173791721-c62b8a53-9d59-4e88-afab-48fb4df24838.png)

Screenshot of certificate manager under trusted publishers:
![MicrosoftTeams-image](https://user-images.githubusercontent.com/16518452/173791753-1de508a3-a399-4c6a-9efe-88bf86822d07.png)

**Comment by fcabrera23** [comment]
Hi @elidevs,

The certificate needed should be delivered through Windows Update.  Are you using an offline version of Windows Server? Could you please apply the Windows updates and try again?

Thanks,
Francisco

**Comment by fcabrera23** [comment]
Hi Eli,

TL;DR - You may have an expired certificate or broken trust chain for your certificates. Run Window Update until it does not offer any additional updates to ensure that your certificates are current and try again.

This condition typically occurs when the trust chain of the certificates on your system is broken or certificates have expired.

1. To check the certificate chain is intact, please go to Start and type "Manage Computer Certificates" and select Open next to the matching application.
2. Locate "Microsoft Code Signing PCA 2011" under Trusted Publishers\Certificates.
3. Check the expiration date
4. Open the certificate and confirm that the description immediately below Certificate Information does not state something like "Windows does not have enough information to verify this certificate."
5. If it does, please run Windows Update to make sure that all updates have been applied to the system. Certificates are regularly updated by Windows Update to prevent this situation.
6. This can also be caused by disabling Automatic Certificate Updates in Group Policy - Additional info, see
http://technet.microsoft.com/en-us/library/cc734054(WS.10).aspx

Thanks,
Francisco

**Comment by elidevs** [comment]
Hi Francisco,

I'm not able to find the "Microsoft Code Signing PCA 2011" certificate in our Azure VM. 

However, I tried deploying EFLOW in my local environment and it succeeded without the "Microsoft Code Signing PCA 2011" certificate.

We are thinking perhaps it is something to do with the firewalls and showing up as a certificate error. Any insights there?

**Comment by fcabrera23** [comment]
Hi @elidevs,

How did you check the certificate?  You should go to "Manage Computer Certificates". The cert should be under **Computer** certificates and not under **User** certificates.

Thanks,
Francisco

**Comment by fcabrera23** [comment]
Hi @elidevs,

Thanks - So you need to have that certificate in order to be able to install EFLOW - If the certificate it's not available, our PSM will try to grab it online and install it. In your case, because it's offline, it won't be able to do that (yes on your local computer). 

Can you try the manual installation of the needed certificates?

1. Download and install Microsoft Code Signing PCA 2011 certificate - [http://www.microsoft.com/pkiops/certs/MicCodSigPCA2011_2011-07-08.crt](https://nam06.safelinks.protection.outlook.com/?url=http%3A%2F%2Fwww.microsoft.com%2Fpkiops%2Fcerts%2FMicCodSigPCA2011_2011-07-08.crt&data=05%7C01%7Cfrancisco.cabrera%40microsoft.com%7Ce7d494765c96447fe27808da5458793e%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C637915036238006772%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=QLenQ5GRXWnPZuePdlnjt4r%2FEg29ZXI%2FrmbQMiB3bdA%3D&reserved=0)
2.  Install Microsoft Corporation  – This is already part of the EFLOW PSM signature; you can extract it and import it before the EFLOW deployment. You can use the following PS code (take into account the EFLOW PSM installation path)
    ```
    $cert = Get-PfxCertificate -FilePath 'C:\Program Files\WindowsPowerShell\Modules\AzureEFLOW\AzureEFLOW.psm1'
    $certStore = Get-Item "cert:\LocalMachine\TrustedPublisher"
    $certStore.Open([System.Security.Cryptography.X509Certificates.OpenFlags]::ReadWrite)
    $certStore.Add($cert)
    $certStore.Close()
    ```

**Comment by fcabrera23** [comment]
Hi @elidevs,

Were you able to do the manual steps for cert installation? Did that work?

Thanks,
Francisco

**Comment by elidevs** [comment]
Hi Francisco,

This issue has been resolved after installing the certificate above.

Thanks for your help!
